docs/architecture/adr-008-square-size-independent-message-commitments.md

```diff
@@ -0,0 +1,123 @@
+
+![Old Commitment](./assets/size-dependent-commitment.png)
+
+To have a commitment independent of the square size, you would need to create a Merkle tree over subtree roots that are included in every possible square size.
+We agree on producing the commitment as if the message was put into the `msgMinSquareSize`. Therefore the subtree roots will be the same and the resulting commitment as well, no matter the square size. *This is because all the subtree roots in the `msgMinSquareSize` are also included in bigger square sizes*.
+You can see this in the updated version of the diagram above. Now the Merkle root will stay the same. This property results from the square width and length being a power of 2. The diagram below has a msgMinSquareSize of 4.
+
+![New Commitment](./assets/size-independent-commitment.png)
+
+You can see in the diagram below that no matter how the message is split up into rows, the hashes of the subtree roots stay the same.
+
+![Row Size Comparison](./assets/row-size-comparison.png)
+
+This scheme also works for interactive commitments as long as the index of when a message starts is the same in every square size. In other words, the leaf nodes and leaf node pairs are independent of the square size.
+For example, in the following diagram, you see the message starting at the second share and still having the same commitment. I marked the skipped blocks pink to show that both messages have the same starting index.
+
+![Interactive Commitment](./assets/interactive-commitment.png)
+
+If the message starting point index is larger than the row of `msgMinSquareSize`, you take the index mod `msgMinSquareSize`. The `msgMinSquareSize` is shown in green, and the skipped block is in pink. In both cases, the commitment stays the same.
+
+![Interactive Commitment 2](./assets/interactive-commitment2.png)
+
+## Alternative Approaches
+
+- [rollmint/adr-007](https://github.com/celestiaorg/rollmint/blob/cb5c7440a8e879778e71097e254c3dd692c39d14/docs/lazy-adr/adr-007-header-commit-to-shares.md)
+
+## Decision
+
+TODO
+
+## Detailed Design
+
+To implement this decision, you need to change [`CreateCommit`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L112-166).
+In Detail, [`powerOf2MountainRange`](https://github.com/celestiaorg/celestia-app/blob/0c81704939cd743937aac2859f3cb5ae6368f174/x/payment/types/payfordata.go#L142) should take `msgMinSquareSize` as an argument instead of `squareSize`.
+
+Following the non-interactive default rules, `msgMinSquareSize` can be calculated like this:
+
+```go
+func msgMinSquareSize(messageLength uint64) uint64 {
+    squareSize := NextHigherPowerOf2(messageLength)
+    //Check if message fits with non-interactive default rules
+    if messageLength <= (squareSize * (squareSize -1)) {
+        return squareSize
+    } else {
+        return squareSize << 1
+    }
+}
+```
```

Comment:
[optional] I think we should calculate msgMinSquareSize without consideration for the current non-interactive defaults. This is safe to do because msgMinSquareSize that ignores non-interactive defaults is less than or equal to a msgMinSquareSize that conforms to current or future non-interactive defaults.

```suggestion
`msgMinSquareSize` can be calculated like this:

```go
// MsgMinSquareSize returns the minimum square size that msgSize can be included
// in. The returned square size does not account for the associated transaction
// shares or non-interactive defaults so it is a minimum.
func MsgMinSquareSize(msgSize uint64) uint64 {
	shareCount := uint64(shares.MsgSharesUsed(int(msgSize)))
	return MinSquareSize(shareCount)
}

// MinSquareSize returns the minimum square size that can contain shareCount
// number of shares.
func MinSquareSize(shareCount uint64) uint64 {
	return shares.RoundUpPowerOfTwo(uint64(math.Ceil(math.Sqrt(float64(shareCount)))))
}
```
```